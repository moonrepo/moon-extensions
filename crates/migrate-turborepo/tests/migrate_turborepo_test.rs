use moon_pdk_test_utils::{create_extension, ExecuteExtensionInput};
use starbase_sandbox::{assert_snapshot, create_sandbox};
use std::fs;

mod migrate_turborepo {
    use super::*;

    #[test]
    fn converts_basic_root_file() {
        let sandbox = create_sandbox("root-only");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });

        assert!(!sandbox.path().join("turbo.json").exists());
        assert!(sandbox.path().join(".moon/tasks/node.yml").exists());

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".moon/tasks/node.yml")).unwrap());
    }

    #[test]
    fn converts_project_files() {
        let sandbox = create_sandbox("monorepo");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });

        assert!(!sandbox.path().join("turbo.json").exists());
        assert!(!sandbox.path().join("client/turbo.json").exists());
        assert!(!sandbox.path().join("server/turbo.json").exists());
        assert!(sandbox.path().join(".moon/tasks/node.yml").exists());
        assert!(sandbox.path().join("client/moon.yml").exists());
        assert!(sandbox.path().join("server/moon.yml").exists());

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".moon/tasks/node.yml")).unwrap());
        assert_snapshot!(fs::read_to_string(sandbox.path().join("client/moon.yml")).unwrap());
        assert_snapshot!(fs::read_to_string(sandbox.path().join("server/moon.yml")).unwrap());
    }

    #[test]
    fn can_force_bun_instead_of_node() {
        let sandbox = create_sandbox("monorepo");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec!["--bun".into()],
            context: plugin.create_context(sandbox.path()),
        });

        assert!(!sandbox.path().join("turbo.json").exists());
        assert!(!sandbox.path().join("client/turbo.json").exists());
        assert!(!sandbox.path().join("server/turbo.json").exists());
        assert!(!sandbox.path().join(".moon/tasks/node.yml").exists());
        assert!(sandbox.path().join(".moon/tasks/bun.yml").exists());
        assert!(sandbox.path().join("client/moon.yml").exists());
        assert!(sandbox.path().join("server/moon.yml").exists());

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".moon/tasks/bun.yml")).unwrap());
        assert_snapshot!(fs::read_to_string(sandbox.path().join("client/moon.yml")).unwrap());
        assert_snapshot!(fs::read_to_string(sandbox.path().join("server/moon.yml")).unwrap());
    }

    #[test]
    fn converts_to_a_root_project() {
        let sandbox = create_sandbox("root-project");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });

        assert!(!sandbox.path().join("turbo.json").exists());
        assert!(!sandbox.path().join(".moon/tasks/node.yml").exists());
        assert!(sandbox.path().join("moon.yml").exists());

        assert_snapshot!(fs::read_to_string(sandbox.path().join("moon.yml")).unwrap());
    }

    #[test]
    fn merges_with_existing_root_tasks() {
        let sandbox = create_sandbox("root-merge-existing");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".moon/tasks/node.yml")).unwrap());
    }

    #[test]
    fn supports_no_pipeline() {
        let sandbox = create_sandbox("missing-pipeline");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });

        assert!(!sandbox.path().join("turbo.json").exists());
        assert!(!sandbox.path().join(".moon/tasks/node.yml").exists());
    }

    #[test]
    #[should_panic(expected = "Unable to migrate task as package client does not exist.")]
    fn errors_if_a_task_points_to_an_unknown_project() {
        let sandbox = create_sandbox("error-missing-project");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });
    }

    #[test]
    #[should_panic(expected = "Unable to migrate task as package client does not exist.")]
    fn errors_if_a_dependson_points_to_an_unknown_project() {
        let sandbox = create_sandbox("error-missing-project-deps");
        let plugin = create_extension("test", sandbox.path());

        plugin.execute_extension(ExecuteExtensionInput {
            args: vec![],
            context: plugin.create_context(sandbox.path()),
        });
    }
}
