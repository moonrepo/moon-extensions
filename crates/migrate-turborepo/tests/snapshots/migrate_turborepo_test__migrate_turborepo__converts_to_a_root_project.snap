---
source: crates/migrate-turborepo/tests/migrate_turborepo_test.rs
expression: "fs::read_to_string(sandbox.path().join(\"moon.yml\")).unwrap()"
---
language: javascript
platform: node
tasks:
  build:
    command: npm run build

